### PR TITLE
perf: optimised to handle dependecies cache

### PR DESCRIPTION
**Pull request by huniafatima-arbi**
This PR includes the strategy for the cache to handle dependencies download.

### CODE REVIEW

File: dockerfiles/registrar.Dockerfile [hunk 1]
```diff
@@ -61,31 +61,42 @@ ENV REGISTRAR_CODE_DIR ${REGISTRAR_CODE_DIR}
...
-
-
 RUN virtualenv -p python${PYTHON_VERSION} --always-copy ${REGISTRAR_VENV_DIR}
 
 RUN pip install --upgrade pip setuptools
 
 ENV REGISTRAR_CFG="${COMMON_CFG_DIR}/registrar.yml"
 
+# extracting requirements folder initially
+RUN curl -L https://github.com/edx/registrar/archive/refs/heads/master/requirements.tar.gz | tar -xz --strip-components=1
```
Comment (mumarkhan999) [inline dockerfiles/registrar.Dockerfile:71]: It's still downloading the complete repo 
image

Comment (huniafatima-arbi) [reply dockerfiles/registrar.Dockerfile]: updated the PR. Please have a look now.

File: dockerfiles/registrar.Dockerfile [hunk 2]
```diff
@@ -77,15 +73,27 @@ EXPOSE 18735
 
 FROM app as dev
 
-RUN pip install --no-cache-dir -r requirements/devstack.txt
+# fetching the requirement file that is needed
+RUN curl -L -o devstack.txt https://raw.githubusercontent.com/edx/registrar/master/requirements/devstack.txt
```
Comment (mumarkhan999) [inline dockerfiles/registrar.Dockerfile:77]:
```suggestion
# fetching the requirement file that is needed
RUN mkdir -p requirements
RUN curl -L -o requirements/devstack.txt https://raw.githubusercontent.com/edx/registrar/master/requirements/devstack.txt
```

File: dockerfiles/registrar.Dockerfile [hunk 3]
```diff
@@ -77,15 +73,27 @@ EXPOSE 18735
...
+
+RUN pip install --no-cache-dir -r ${REGISTRAR_CODE_DIR}/devstack.txt
```
Comment (mumarkhan999) [inline dockerfiles/registrar.Dockerfile:79]:
```suggestion
RUN pip install --no-cache-dir -r ${REGISTRAR_CODE_DIR}/requirements/devstack.txt
```

File: dockerfiles/registrar.Dockerfile [hunk 4]
```diff
@@ -77,15 +73,27 @@ EXPOSE 18735
...
 
 ENV DJANGO_SETTINGS_MODULE registrar.settings.devstack
 
 CMD while true; do python ./manage.py runserver 0.0.0.0:18734; sleep 2; done
 
 FROM app as prod
 
-RUN pip install  --no-cache-dir -r ${REGISTRAR_CODE_DIR}/requirements/production.txt
+# fetching the requirement file that is needed
+RUN curl -L -o production.txt https://raw.githubusercontent.com/edx/registrar/master/requirements/production.txt
```
Comment (mumarkhan999) [inline dockerfiles/registrar.Dockerfile:91]:
```suggestion
RUN curl -L -o requirements/production.txt https://raw.githubusercontent.com/edx/registrar/master/requirements/production.txt
```

File: dockerfiles/registrar.Dockerfile [hunk 5]
```diff
@@ -77,15 +73,27 @@ EXPOSE 18735
...
+
+RUN pip install --no-cache-dir -r ${REGISTRAR_CODE_DIR}/production.txt
```
Comment (mumarkhan999) [inline dockerfiles/registrar.Dockerfile:93]:
```suggestion
RUN pip install --no-cache-dir -r ${REGISTRAR_CODE_DIR}/requirements/production.txt
```

File: dockerfiles/registrar.Dockerfile [hunk 6]
```diff
@@ -61,10 +61,6 @@ ENV REGISTRAR_CODE_DIR ${REGISTRAR_CODE_DIR}
 # Working directory will be root of repo.
 WORKDIR ${REGISTRAR_CODE_DIR}
```
Comment (mumarkhan999) [inline dockerfiles/registrar.Dockerfile:62]:
```suggestion
WORKDIR ${REGISTRAR_CODE_DIR}
# Create requirements directory
RUN mkdir -p requirements
```

**Review by UsamaSadiq — APPROVED** [review]

**Pull request merged**